builtin/providers/aws/resource_aws_instance.go

```diff
@@ -923,17 +935,47 @@ type awsInstanceOpts struct {
 	UserData64                        *string
 }
 
-func buildAwsInstanceOpts(
-	d *schema.ResourceData, meta interface{}) (*awsInstanceOpts, error) {
+func getAwsAmiId(d *schema.ResourceData, conn *ec2.EC2) string {
+	ami_id := d.Get("ami").(string)
+	ami_name := d.Get("ami_name").(string)
+
+	if ami_id != "" {
+		return ami_id
+	} else if ami_name != "" {
+		opts := &ec2.DescribeImagesInput{
+			Filters: []*ec2.Filter{&ec2.Filter{
+				Name:   aws.String("name"),
+				Values: []*string{aws.String(d.Get("ami_name").(string))},
+			}},
+		}
+		log.Printf("[INFO] Trying to find the AMI ID for the AMI Name: %s", ami_name)
+		images, err := conn.DescribeImages(opts)
+		if err != nil {
+			log.Printf("Error Finding The ImageId for AMI Name: %s : %s", ami_name, err)
+		}
+
+		return *images.Images[0].ImageId
```

Comment:
We should probably range through the results returned and only return the exact match. I've seen other situations where `name` and `name1` both match a `Describe` call and `[0]` is not guaranteed to be the one you want